File: ED_RRiveraG1_A09/MyDLList.cpp
//
// Created by Roberto Rivera on 9/21/17.
//

#include <iostream>
#include "NodeDL.cpp"

template <typename T> class MyDLList {
public:
    NodeDL<T> *first;
    NodeDL<T> *last;
    NodeDL<T> *pointer;
    int length;

    MyDLList() {
        first = nullptr;
        last = nullptr;
        pointer = nullptr;
        length = 0;
    }

    MyDLList(T d) {
        NodeDL<T> *node = new NodeDL<T>(d);
        *first = node;
        *last = node;
        pointer = nullptr;
        length = 1;
    }

    bool isEmpty() {
        return length == 0;
    }

    void insertFirst(T d) {
        NodeDL<T> *node = new NodeDL<T>(d);
        if(isEmpty()){
            first = node;
            last = node;
        } else {
            node->next = first;
            first->prev = node;
            first = node;
        }
        length+=1;
    }
    void insertLast(T d){
        NodeDL<T> *node = new NodeDL<T>(d);
        if(!isEmpty()){
            last->next=node;
            node->prev=last;
            last=node;
        }else{
            first=node;
            last=node;
        }
        length+=1;
    }
    void borrarInicio(){
        if(!isEmpty()){
            first->next->prev= nullptr;
            first=first->next;
            length-=1;
        }
    }
    void borrarFinal(){
        if(!isEmpty()){
            last=last->prev;
            last->next= nullptr;
            length-=1;
        }
    }
    void borrarNodo(T data){
        if(!isEmpty()){
            if(first->data==data){
                borrarInicio();
                length-=1;
                return;
            }
            if(last->data==data){
                borrarFinal();
                length-=1;
                return;
            }
            pointer=first;
            do{
                if(pointer->data==data){
                    pointer->prev->next=pointer->next;
                    pointer->next->prev=pointer->prev;
                    pointer= nullptr;
                    length-=1;
                } else pointer=pointer->next;
            } while(pointer!= nullptr);
        }
    }
    void muestraListaInicioFin(){
        if(!isEmpty()){
            pointer=first;
            std::cout << "First->";
            while(pointer!= nullptr){
                std::cout << "[" << pointer->data << "]";
                pointer=pointer->next;
                if(pointer!= nullptr) std::cout << " -> ";
            }
            std::cout << " <- Last"<< std::endl;
        }
    }

    void muestraListaFinInicio(){
        if(!isEmpty()){
            pointer=last;
            std::cout << "Last->";
            while(pointer!= nullptr){
                std::cout << "[" << pointer->data << "]";
                pointer=pointer->prev;
                if(pointer!= nullptr) std::cout << " -> ";
            }
            std::cout << " <- First"<< std::endl;
        }
    }
};

File: Listas/Java/DoubleLinkedList/MyDLList.java
package ED.Riverto;

public class MyDLList <T> {
    NodeDL<T> first;
    NodeDL<T> last;
    NodeDL<T> pointer;
    int length;

    public MyDLList() {
        first = null;
        last = null;
        pointer = null;
        length = 0;
    }
    public MyDLList(T d) {
        NodeDL<T> node = new NodeDL<T>(d);
        first = null;
        last = node;
        pointer = null;
        length = 1;
    }

    public boolean isEmpty() {
        return length == 0;
    }

    public void insertFirst(T d) {
        NodeDL<T> node = new NodeDL<T>(d);
        if(isEmpty()){
            first = node;
            last = node;
        } else {
            node.next = first;
            first.prev = node;
            first = node;
        }
        length+=1;
    }
    public void insertLast(T d){
        NodeDL<T> node = new NodeDL<T>(d);
        if(!isEmpty()){
            last.next=node;
            node.prev=last;
            last=node;
        }else{
            first=node;
            last=node;
        }
        length+=1;
    }
    public void borrarInicio(){
        if(!isEmpty()){
            first.next.prev= null;
            first=first.next;
            length-=1;
        }
    }
    public void borrarFinal(){
        if(!isEmpty()){
            last=last.prev;
            last.next= null;
            length-=1;
        }
    }
    public void borrarNodo(T data){
        if(!isEmpty()){
            if(first.data==data){
                borrarInicio();
                length-=1;
                return;
            }
            if(last.data==data){
                borrarFinal();
                length-=1;
                return;
            }
            pointer=first;
            do{
                if(pointer.data==data){
                    pointer.prev.next=pointer.next;
                    pointer.next.prev=pointer.prev;
                    pointer= null;
                    length-=1;
                } else pointer=pointer.next;
            } while(pointer!= null);
        }
    }
    public void muestraListaInicioFin(){
        if(!isEmpty()){
            pointer=first;
            System.out.print("First->");
            while(pointer!= null){
                System.out.print("[" + pointer.data + "]");
                pointer=pointer.next;
                if(pointer!= null) System.out.print(" <=> ");
            }
            System.out.print(" <- Last \n");
        }
    }

    public void muestraListaFinInicio(){
        if(!isEmpty()){
            pointer=last;
            System.out.print("Last->");
            while(pointer!= null){
                System.out.print("[" + pointer.data + "]");
                pointer=pointer.prev;
                if(pointer!= null) System.out.print(" <=> ");
            }
            System.out.print(" <- First\n");
        }
    }
}
